includes/User.php

```diff
@@ -5237,4 +4613,256 @@ protected function logFallbackToMediaWikiSessionRejection( $from ) {
+		return self::userPermissions()->getEffectiveUserGroups( $this, $recache );
+	}
+
+	/**
+	 * Get the list of implicit group memberships this user has.
+	 * This includes 'user' if logged in, '*' for all accounts,
+	 * and autopromoted groups
+	 * @param $recache Bool Whether to avoid the cache
+	 * @return Array of String internal group names
+	 */
+	public function getAutomaticGroups( $recache = false ) {
+		return self::userPermissions()->getAutomaticUserGroups( $this, $recache );
+	}
+
+	/**
+	 * Get a list of implicit groups
+	 * @return Array of Strings Array of internal group names
+	 */
+	public static function getImplicitGroups() {
+		return self::userPermissions()->getImplicitGroups();
+	}
+
+	/**
+	 * Return the set of defined explicit groups.
+	 * The implicit groups (by default *, 'user' and 'autoconfirmed')
+	 * are not included, as they are defined automatically, not in the database.
+	 * @return Array of internal group names
+	 */
+	public static function getAllGroups() {
+		return self::userPermissions()->getExplicitGroups();
+	}
+
+	/**
+	 * Get the permissions associated with a given list of groups
+	 *
+	 * @param $groups Array of Strings List of internal group names
+	 * @return Array of Strings List of permission key names for given groups combined
+	 */
+	public static function getGroupPermissions( $groups ) {
+		return self::userPermissions()->getGroupPermissions( $groups );
+	}
+
+	/**
+	 * Get all the groups who have a given permission
+	 *
+	 * @param $role String Role to check
+	 * @return Array of Strings List of internal group names with the given permission
+	 */
+	public static function getGroupsWithPermission( $role ) {
+		return self::userPermissions()->getGroupsWithPermission( $role );
+	}
+
+	/**
+	 * Get the permissions this user has.
+	 * @return Array of String permission names
+	 */
+	public function getRights() {
+		return self::userPermissions()->getUserPermissions( $this );
+	}
+
+	/**
+	 * Get a list of all available permissions.
+	 * @return Array of permission names
+	 */
+	public static function getAllRights() {
+		return self::userPermissions()->getPermissions();
+	}
+
+	/**
+	 * Returns an array of the groups that a particular group can add/remove.
+	 *
+	 * @param $group String: the group to check for whether it can add/remove
+	 * @return Array array( 'add' => array( addablegroups ),
+	 *     'remove' => array( removablegroups ),
+	 *     'add-self' => array( addablegroups to self),
+	 *     'remove-self' => array( removable groups from self) )
+	 */
+	public static function changeableByGroup( $group ) {
+		return self::userPermissions()->getGroupsChangeableByGroup( $group );
+	}
+
+	/**
+	 * Add the user to the given group.
+	 * This takes immediate effect.
+	 * @param $group String Name of the group to add
+	 */
+	public function addGroup( $group ) {
+		global $wgUser;
+		return self::userPermissions()->addUserToGroup( $wgUser, $this, $group );
```

Comment:
Ok, I can change that, but you could explain why using wgUser is bad? Just because we want to avoid globals in general is there some timing issue? I'm asking because I use wgUser in other places as well